// @aragon/osx-commons-configs/dist/deployments/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contracts = exports.zksyncMainnet = exports.zksyncSepolia = exports.arbitrumSepolia = exports.arbitrum = exports.baseSepolia = exports.baseGoerli = exports.baseMainnet = exports.mumbai = exports.polygon = exports.sepolia = exports.goerli = exports.mainnet = void 0;
const tslib_1 = require("tslib");
const arbitrum = tslib_1.__importStar(require("./json/arbitrum.json"));
exports.arbitrum = arbitrum;
const arbitrumSepolia = tslib_1.__importStar(require("./json/arbitrumSepolia.json"));
exports.arbitrumSepolia = arbitrumSepolia;
const baseGoerli = tslib_1.__importStar(require("./json/baseGoerli.json"));
exports.baseGoerli = baseGoerli;
const baseMainnet = tslib_1.__importStar(require("./json/baseMainnet.json"));
exports.baseMainnet = baseMainnet;
const baseSepolia = tslib_1.__importStar(require("./json/baseSepolia.json"));
exports.baseSepolia = baseSepolia;
const goerli = tslib_1.__importStar(require("./json/goerli.json"));
exports.goerli = goerli;
const mainnet = tslib_1.__importStar(require("./json/mainnet.json"));
exports.mainnet = mainnet;
const mumbai = tslib_1.__importStar(require("./json/mumbai.json"));
exports.mumbai = mumbai;
const polygon = tslib_1.__importStar(require("./json/polygon.json"));
exports.polygon = polygon;
const sepolia = tslib_1.__importStar(require("./json/sepolia.json"));
exports.sepolia = sepolia;
const zksyncMainnet = tslib_1.__importStar(require("./json/zksyncMainnet.json"));
exports.zksyncMainnet = zksyncMainnet;
const zksyncSepolia = tslib_1.__importStar(require("./json/zksyncSepolia.json"));
exports.zksyncSepolia = zksyncSepolia;
const types_1 = require("./types");
exports.contracts = {
    mainnet,
    goerli,
    sepolia,
    polygon,
    mumbai,
    baseMainnet,
    baseGoerli,
    baseSepolia,
    arbitrum,
    arbitrumSepolia,
    zksyncSepolia,
    zksyncMainnet,
    local: {
        [types_1.SupportedVersions.V1_0_0]: {},
        [types_1.SupportedVersions.V1_3_0]: {},
    },
};
