// @aragon/osx-commons-configs/dist/networks/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedNetworks = exports.SupportedAliases = void 0;
var SupportedAliases;
(function (SupportedAliases) {
    SupportedAliases["ETHERS_5"] = "ethers5";
    SupportedAliases["ETHERS_6"] = "ethers6";
    SupportedAliases["ALCHEMY_SUBGRAPHS"] = "alchemySubgraphs";
})(SupportedAliases || (exports.SupportedAliases = SupportedAliases = {}));
var SupportedNetworks;
(function (SupportedNetworks) {
    SupportedNetworks["MAINNET"] = "mainnet";
    SupportedNetworks["GOERLI"] = "goerli";
    SupportedNetworks["SEPOLIA"] = "sepolia";
    SupportedNetworks["POLYGON"] = "polygon";
    SupportedNetworks["MUMBAI"] = "mumbai";
    SupportedNetworks["BASE"] = "baseMainnet";
    SupportedNetworks["BASE_GOERLI"] = "baseGoerli";
    SupportedNetworks["BASE_SEPOLIA"] = "baseSepolia";
    SupportedNetworks["ARBITRUM"] = "arbitrum";
    SupportedNetworks["ARBITRUM_SEPOLIA"] = "arbitrumSepolia";
    SupportedNetworks["ZKSYNC_SEPOLIA"] = "zksyncSepolia";
    SupportedNetworks["ZKSYNC_MAINNET"] = "zksyncMainnet";
    SupportedNetworks["LOCAL"] = "local";
})(SupportedNetworks || (exports.SupportedNetworks = SupportedNetworks = {}));
